### ISSUE TITLE
Further ranged balance tests.

### ISSUE THREAD

**Issue by Firestorm01X2** [CleverRaven/Cataclysm-DDA#22211]
Continuation of:
https://github.com/CleverRaven/Cataclysm-DDA/issues/22214

Current ranged tests doesn't satisfy balance design outlined in:
https://github.com/CleverRaven/Cataclysm-DDA/issues/21244#issuecomment-310013403
or it's modified version.

Current problems:
1) Balance tables in design based on *good_hit* while tests check *normal_hit*. That leads to lower actual accuracy.
2) Tests doesn't cover shotguns.
3) Test characters do not fit actual ideas of accuracy levels outlined in design. They are too "debug" charaters but not "actual" characters that player may use.
4) Tests were regressed as mentioned in: https://github.com/CleverRaven/Cataclysm-DDA/pull/22146
Tests were tweaked to fit themself.
4) Neded additional tests for long range sniping for competent shooter. For example:
Based on Coolthulhu's design:

Character | >50% good | <10% good
-- | -- | --
Competent rifle | 15 | **45**

6) Needed tests between >50% and <10% ranges. Currently tested min_good_range and max_good_range fo each case. There are must be tests in between two ranges.

So competent sniper should be able to hit something stationary on 30 tiles with around 30% chance.  Good test character for that test is Police Sniper class.

Tests must satisfy design idea: https://github.com/CleverRaven/Cataclysm-DDA/issues/21244#issuecomment-310013403
or it's modified version.

Important note- this issue only for tests themself. It is not for current balance state.

Do not recommended actions:
1) Force test to fit themself. If it would be done again - then we will return to same issue again. And again.

**Comment by Firestorm01X2** [comment]
If ranged balance design is not accepted - than it mus be discussed first!

**Comment by kevingranade** [comment]
> this issue only for tests but not for actual balance
In particular, I don't know where you get this idea:
> Let tests fail until proper balance would be achieved

No changes to code or tests that cause there to be failing tests should be pushed to the main branch of the repository.  Changes to tests that cause them to fail must be accompanied by changes that make them pass again.  This issue is very much about actual game balance, not just adjusting tests.

> If ranged balance design is not accepted - than it must be discussed first!

It was discussed extensively in #21244, all departures from the suggested tests were intentional and documented, I don't like your implication that it wasn't.

**Comment by Firestorm01X2** [comment]
>It was discussed extensively in #21244, all departures from the suggested tests were intentional and documented, I don't like your implication that it wasn't.

Well, sorry sometimes it is hard to follow what was accepted and what was not. Especially if devs consult each other outside github.

>No changes to code or tests that cause there to be failing tests should be pushed to the main branch of the repository. Changes to tests that cause them to fail must be accompanied by changes that make them pass again. This issue is very much about actual game balance, not just adjusting tests.

It is actually do not referred to my PR. It is to idead itself. Create test that satisfy design. Approve it. Push it and then step by step achieve it.

**Comment by kevingranade** [comment]
> But you mentioned that we can't allow multiple branches online because there are not enough developers to mantain multiple online branches.

We can in fact have a feature branch for ranged balance with incremental pushes that is allowed to fail.  If that is what you mean, then it's my mistake and what you're suggesting is fine.

If there is only one contributor pushing changes to this branch, it is no different than a PR branch (they actually do have branches, but they are generally short-lived and exist only in a user repository until they are pushed).  But we can also have a series of pull requests by different contributors that are merged to such a feature branch if that is desired.

> It is important to have proper tests.

The existing tests are "proper", they make valid assertions about the limits of ranged accuracy.  Further tests can increase coverage and tighten the guarantees, but that makes them *incomplete* not *improper*.

> It missed few design points

On this point I do agree, there is additional room for improvement in the tests:
> Balance tables in design based on good_hit while tests check normal_hit. That leads to lower actual accuracy.

I disagree this is an issue, Coolthulu's tables were suggested guidelines, not precisely-tuned rules.  The adjustments to them I made were intentional in order to rationalize some of them and account for additional requirements.
> Tests doesn't cover shotguns.

Agreed, we need some adjustment to the goals and additional test cases to cover shotguns, as they have a unique niche within firearms.

> Test characters do not fit actual ideas of accuracy levels outlined in design. They are too "debug" charaters but not "actual" characters that player may use.

This is the most serious issue IMO.  This will require coming up with some plan for adjusting the impact of weapon mods, possibly extensively, and applying them.  Most likely this will not change the tests themselves, but instead just adjust the NPC gear lists used, as well as adjusting the stats of that gear.

> Tests were regressed as mentioned in: #22146
>   Tests were tweaked to fit themself.

Disagree, I never agreed this Dex/Per thing was a valid issue.  For one thing there's no definition of what is required to address it, and for another there's no actual requirement for stats to be "equal".

**Comment by Firestorm01X2** [comment]
>I disagree this is an issue, Coolthulu's tables were suggested guidelines, not precisely-tuned rules. The adjustments to them I made were intentional in order to rationalize some of them and account for additional requirements.

What is what for first issue were about. So there is still room to discuss in design. Anyway. Using normal_hit for that tables interfere with Coolthulhu idea and make overall accuracy lower. That is exactly what happened now. I saw player reports here and there. They said in short: "Accuracy is better but overall it is still not enough."

**Comment by Firestorm01X2** [comment]
So this issue will stay. Also that aiming for good_hit Is in Collthulhu design. It is better go for that because normal_hit does lower damage then weapon damage.

**Comment by Firestorm01X2** [comment]
I want to mention one more thing. 
There is must be test checking ability of competent shooter to hit something on range 30 with appropriate sniper rifle. 
Based on Coolthulhu's design:

Competent rifle | 15 | **45**
-- | -- | --

So competent sniper should be able to hit something stationary on 30 tile with around 30% chance. Now it is hardly the case. Good test character for that is Police sniper.

**Comment by Firestorm01X2** [comment]
This issue likely to be ignored. So it is closed.

**Comment by Night-Pryanik** [comment]
Should be closed after the ranged balance PRs?

**Comment by kevingranade** [comment]
The last remaining issue is captured by #24429
